README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [mnubo Java SDK](#mnubo-java-sdk)
- [Introduction](#introduction)
- [Architecture](#architecture)
- [Prerequisites](#prerequisites)
- [Installation & Configuration](#installation-&-configuration)
  - [Maven](#maven)
  - [Download source code](#download-source-code)
    - [Configuration](#configuration)
- [Usage](#usage)
  - [Getting a "MnuboSDKClient" (client) instance](#getting-a-mnubosdkclient-client-instance)
    - [Basic Configuration](#basic-configuration)
    - [Advanced Configuration](#advanced-configuration)
  - [Creating owners](#creating-owners)
  - [Updating owners](#updating-owners)
  - [Create or update a batch of Owners](#create-or-update-a-batch-of-owners)
  - [Deleting owners](#deleting-owners)
  - [Check if an owner exists](#check-if-an-owner-exists)
  - [Check if a batch of owners exist](#check-if-a-batch-of-owners-exist)
  - [Claiming an object](#claiming-an-object)
  - [Creating objects](#creating-objects)
  - [Updating objects](#updating-objects)
  - [Create or update a batch of Objects](#create-or-update-a-batch-of-objects)
  - [Deleting objects](#deleting-objects)
  - [Check if an object exists](#check-if-an-object-exists)
  - [Check if a batch of objects exist](#check-if-a-batch-of-objects-exist)
  - [Sending Events](#sending-events)
    - [To send multiple events to a single object:](#to-send-multiple-events-to-a-single-object)
    - [To send multiple events to multiple objects:](#to-send-multiple-events-to-multiple-objects)
  - [Check if an event exists](#check-if-an-event-exists)
  - [Check if a batch of events exist](#check-if-a-batch-of-events-exist)
  - [Searching](#searching)
    - [Sending search query](#sending-search-query)
  - [Retrieving datasets](#retrieving-datasets)
- [References](#references)
- [Configuring the example](#configuring-the-example)
  - [Running the example](#running-the-example)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# mnubo Java SDK

[![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.mnubo/java-sdk-client/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.mnubo/java-sdk-client)

Introduction
============
This is a Java implementation of the [API documentation](https://sop.mtl.mnubo.com/apps/doc/?i=t).

Architecture
============
Use class `MnuboSDKFactory` to get a client instance for ObjectsSDK, OwnersSDK, EventsSDK or SearchSDK.  Then the appropriate client API.

The current SDK only supports synchronous calls.


Prerequisites
=============

- Using this java SDK requires JAVA 1.7 or higher.
- The SDK has been built with maven 3.2.3.


Installation & Configuration
============================

Include the mnubo client in your Java application using:

Maven
-----
You can find information about dependencies to add into your Pom.xml here.

https://maven-badges.herokuapp.com/maven-central/com.mnubo/java-sdk-client

Download source code
---------------------
Download the source code and include it in your Java Application project.

### Configuration
The following parameters must be configured before using the mnubo client:

    - **hostname**.- mnubo's server name, for example: ```rest.sandbox.mnubo.com```.
    - **consumer-key**.- Your unique client identity which is provided by mnubo.
    - **consumer-secret**.- Your secret key which is used in conjunction with the consumer key to access the mnubo server. This key is provided by mnubo.


Usage
=====

Getting a "MnuboSDKClient" (client) instance
--------------------------------------------
To get a client instance use the **"MnuboSDKFactory" Class**. Note that you only need one client instance. We provide multithread support and a pool of connections.

There are two ways to obtain a client instance:

### Basic Configuration
Using the basic method, three mandatory parameters are required – Host, consumer key and consumer secret. Please see the example below:


```
//Example:
//Configure constants
private final String HOST = "rest.sandbox.mnubo.com";
private final String CONSUMER_KEY = "your consumer key!!!";
private final String CONSUMER_SECRET = "your consumer SECRET!!!";

//Obtain a client instance using default values.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );
```

### Advanced Configuration
Using this method you configure mandatory and optional parameters in a properties file. If you wish to use this method please contact mnubo.

Creating owners
---------------
To create an owner you need to:
1. Request an OwnersSDK interface using the mnubo client instance.
2. Build an owner.

This example describes how to create an Owner:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//build the owner
Owner Owner2Create = Owner.builder()
                          .withUsername("john.smith@mycompany.com")
                          .withPassword("dud7#%^ddd_J")
                          .withRegistrationDate(DateTime.parse("2015-01-01T00:00:00+04:00"))
                          .withAddedAttribute("age", 35)
                          .withAddedAttribute("gender", "male")
                          .withAddedAttribute("height", 1.80)
                          .build();

//create the owner
mnuboOwnersClient.create( Owner2Create );
```

You can also create an owner using a JSON file:

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//read the Owner from the JSON file
String owner2BePosted = readingFile( "myOwnerFile.json" );

//deserialise the JSON file
Owner owner2Create = SDKMapperUtils.readValue( owner2BePosted , Owner.class );

//create the owner
mnuboOwnersClient.create( owner2Create );
```

The JSON file "myOwnerFile.json" is as follows:

```
{
  "username":"john.smith@mycompany.com",
  "x_password":"dud7#%^ddd_J",
  "x_registration_date":"2015-01-01T00:00:00+04:00",
  "age":35,
  "gender":"male",
  "height":1.80
}
```

Updating owners
---------------
To update an owner you need to:
1. Request an OwnersSDK interface using the mnubo client instance.
2. Build an owner with the parameters to change.

This example describes how to update an Owner:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//build the owner, including only the parameters to update. Remember "username" & "password" cannot change
Owner Owner2Update = Owner.builder()
                          .withRegistrationDate(DateTime.parse("2015-01-02T00:00:00+04:00"))
                          .withAddedAttribute("age", 38)
                          .build();

//update the owner
mnuboOwnersClient.update( Owner2Update, "john.smith@mycompany.com" );
```

You can also update an owner using a JSON file:

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//read the Owner from the JSON file
String owner2BePosted = readingFile( "myOwnerFile.json" );

//deserialise the JSON file
Owner owner2Update = SDKMapperUtils.readValue( owner2BePosted , Owner.class );

//update the owner
mnuboOwnersClient.update( owner2Update, "john.smith@mycompany.com" );
```

The JSON file "myOwnerFile.json" is as follows. remember only parameters to change:

```
{
  "x_registration_date":"2015-01-02T00:00:00+04:00",
  "age":38,
}
```

Create or update a batch of Owners
---------------
To create a batch of owners you need to:
1. Request an OwnersSDK interface using the mnubo client instance.
2. Build a list of Owners.

This example describes how to create a batch of Owners:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//build the list of owner to create / update
List<Owner> owners = new ArrayList<>();

owners.add(Owner.builder()
                .withUsername("user1@mycompany.com")
                .withPassword("dud7#%^ddd_J")
                .withRegistrationDate(DateTime.parse("2015-01-01T00:00:00+04:00"))
                .withAddedAttribute("age", 35)
                .withAddedAttribute("gender", "male")
                .withAddedAttribute("height", 1.80)
                .build());

owners.add(Owner.builder()
                .withUsername("user2@mycompany.com")
                .withPassword("ewor$@$!@fqrq")
                .withRegistrationDate(DateTime.parse("2015-02-01T00:00:00+04:00"))
                .withAddedAttribute("age", 18)
                .withAddedAttribute("gender", "male")
                .withAddedAttribute("height", 1.75)
                .build());

// A single batch can contain up to 1000 Owners

//create the list of owners
List<Result> results = mnuboOwnersClient.createUpdate( owners );
```

You can verify if there is any error(s) for the list of owners created / updated.

The API result will contain a list of the username's (id), their results (“success” or “error”)
and a message if an error occurred.

Note: ResultStates is an ENUM with only ResultStates.success (“success”) or ResultStates.error (“error”)

Here is an example to get the data of the list of Result:
```
for(Result r: results) {
    String id = r.getId();
    ResultStates resultState = r.getResult();
    String message = r.getMessage();
}
```

Deleting owners
---------------
This example describes how to delete an Owner:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//delete the owner
mnuboOwnersClient.delete( "john.smith@mycompany.com" );
```

Check if an owner exists
------------------------
This example describes how to validate if an Owner exists:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//true if the owner exists, false if not.
boolean doesExists = mnuboOwnersClient.ownerExists( "john.smith@mycompany.com" );
```

Check if a batch of owners exist
--------------------------------
This example describes how to validate if a list of Owners exists:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

Map<String, Boolean> ownersExist = mnuboOwnersClient.ownersExist( Arrays.asList("john.smith@mycompany.com", "my.test@mycompany.com") );
//true if the owner `john.smith@mycompany.com` exists, false if not.
boolean johnSmithExists = ownersExist.get("john.smith@mycompany.com")
//true if the owner `my.test@mycompany.com` exists, false if not.
boolean myTestExists = ownersExist.get("my.test@mycompany.com")
```

Claiming an object
------------------
This example describes how to link an Object with an Owner:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Owners client interface
OwnersSDK mnuboOwnersClient = mnuboClient.getOwnerClient();

//claim the object
mnuboOwnersClient.claim( "john.smith@mycompany.com", "my_device_Id" );
```

Creating objects
----------------------
To create an object:
1. Request an ObjectSDK interface from the mnubo client instance.
2. Build an object.

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//build the object
SmartObject object2Create = SmartObject.builder()
                                       .withDeviceId("connect_alpha.6hv135nw00393.1234567")
                                       .withObjectType("gateway")
                                       .withOwner("john.smith@mycompany.com")
                                       .withRegistrationDate(DateTime.now())
                                       .withAddedAttribute("partnerid", "connect_alpha")
                                       .withAddedAttribute("business_line", "connect")
                                       .withAddedAttribute("siteid", "6hv135nw00393")
                                       .withAddedAttribute("site_description", "My connected House")
                                       .build();

//create the object
mnuboObjectClient.create( object2Create );
```

You can also create an object using a JSON file

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//read the Object from the JSON file
String object2BePosted = readingFile( "myObjectFile.json" );

//deserialise the JSON file
SmartObject object2Create = SDKMapperUtils.readValue( object2BePosted , SmartObject.class );

//create the object
mnuboObjectClient.create( object2Create );
```

The JSON file "myObjectFile.json" is as follows:

```
{
    "x_device_id" : "connect_alpha.6hv135nw00393.1234567",
    "x_object_type" : "gateway",
    "x_registration_date":"2015-01-27T08:01:01.000Z",
    "partnerid" : "connect_alpha",
    "business_line" : "connect",
    "siteid" : "6hv135nw00393",
    "site_description" : "My connected House",
    "x_owner":
    {
           "username" : "john.smith@mycompany.com"
    }
}
```

Updating objects
----------------
To update an object:
1. Request an ObjectSDK interface from the mnubo client instance.
2. Build an object with parameters to change.

Note: that an owner cannot be added when updating an object.

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//build the object, only with parameters to change.
SmartObject object2Update = SmartObject.builder()
                                       .withDeviceId("connect_beta.6hv135nw00393.1234567")
                                       .withObjectType("gateway1")
                                       .withAddedAttribute("site_description", "My connected building")
                                       .build();

//update the object
mnuboObjectClient.update( object2Update, "connect_alpha.6hv135nw00393.1234567" );
```

You can also update an object using a JSON file

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//read the Object from the JSON file
String object2BePosted = readingFile( "myObjectFile.json" );

//deserialise the JSON file
SmartObject object2Update = SDKMapperUtils.readValue( object2BePosted , SmartObject.class );

//update the object
mnuboObjectClient.update( object2Update, "connect_alpha.6hv135nw00393.1234567" );
```

The JSON file "myObjectFile.json" is as follows:

```
{
    "x_device_id" : "connect_beta.6hv135nw00393.1234567",
    "x_object_type" : "gateway1",
    "site_description" : "My connected building",
}
```

Create or update a batch of Objects
-----------------------------------
To create a batch of objects you need to:
1. Request an ObjectSDK interface from the mnubo client instance.
2. Build a list of objects.

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//build the list of objects to create / update

List<SmartObject> objects = new ArrayList<>();

objects.add(SmartObject.builder()
                       .withDeviceId("connect_alpha.6hv135nw00393.1234567")
                       .withObjectType("gateway")
                       .withOwner("user1@mycompany.com")
                       .withRegistrationDate(DateTime.now())
                       .withAddedAttribute("partnerid", "connect_alpha")
                       .withAddedAttribute("business_line", "connect")
                       .withAddedAttribute("siteid", "6hv135nw00393")
                       .withAddedAttribute("site_description", "My connected House 1")
                       .build());

objects.add(SmartObject.builder()
                       .withDeviceId("connect_beta.7hv234nw60696.7654321")
                       .withObjectType("gateway")
                       .withOwner("user2@mycompany.com")
                       .withRegistrationDate(DateTime.now())
                       .withAddedAttribute("partnerid", "connect_beta")
                       .withAddedAttribute("business_line", "connect")
                       .withAddedAttribute("siteid", "7hv234nw60696")
                       .withAddedAttribute("site_description", "My connected House 2")
                       .build());

// A single batch can contain up to 1000 Objects

// Create the list of objects
List<Result> results = mnuboObjectClient.createUpdate( objects );

```
You can verify if there is any error(s) for the list of owners created / updated.

The API result will contain a list of the x_device_id's (id), their results (“success” or “error”)
and a message if an error occurreds.

Note: ResultStates is an ENUM with only ResultStates.success (“success”) or ResultStates.error (“error”)

Here is an example to get the data of the list of Result:
```
for(Result r: results) {
    String id = r.getId();
    ResultStates resultState = r.getResult();
    String message = r.getMessage();
}
```

Deleting objects
----------------
```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//delete the object
mnuboObjectClient.delete( "connect_alpha.6hv135nw00393.1234567" );
```

Check if an object exists
-------------------------
This example describes how to validate if an Object exists:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

//true if the object exists, false if not.
boolean doesExists = mnuboObjectClient.objectExists( "connect_alpha.6hv135nw00393.1234567" );
```

Check if a batch of objects exist
---------------
This example describes how to validate if a list of Objects exists:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Object client interface
ObjectsSDK mnuboObjectClient = mnuboClient.getObjectClient();

Map<String, Boolean> objectsExist = mnuboObjectClient.objectsExist( Arrays.asList("connect_alpha.6hv135nw00393.1234567", "connect_beta.1234567") );
//true if the device `connect_alpha.6hv135nw00393.1234567` exists, false if not.
boolean deviceAExists = objectsExist.get("connect_alpha.6hv135nw00393.1234567")
//true if the device `connect_beta.1234567` exists, false if not.
Boolean deviceBExists = objectsExist.get("connect_beta.1234567")
```

Sending Events
--------------
To send events:
1. Request an OwnersSDK interface from the mnubo client instance.
2. Build an event.

##### To send multiple events to a single object:

```
//private String objectID = "mythermostat0301424";

//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Event client interface
EventsSDK mnuboEventClient = mnuboClient.getEventClient();

//build the events
List<Event> event2Send = new ArrayList<Event>();
Event event1 = Event.builder()
                    .withEventType("thermostat_temperature")
                    .withEventID(UUID.fromString("aac652a3-955e-4ac4-b185-b7c2f44111cc"))
                    .withTimestamp(DateTime.now())
                    .withAddedTimeseries("temperature", 20)
                    .withAddedTimeseries("errorcode", "")
                    .withAddedTimeseries("varname", "temperature")
                    .build();
event2Send.add(event1);
Event event2 = Event.builder()
                    .withEventType("thermostat_temperature")
                    .withEventID(UUID.fromString("583acbba-8a4f-4503-ac49-bc5a8e4f7577"))
                    .withTimestamp(DateTime.now())
                    .withAddedTimeseries("temperature", 22)
                    .withAddedTimeseries("cause_type", "normal")
                    .withAddedTimeseries("varname", "temperature")
                    .build();
event2Send.add(event2);

//send the event
List<Result> results = mnuboEventClient.send( objectID, event2Send );
```

You can also send multiple events to a single object using a JSON file:

```
//private String objectID = "mythermostat0301424";

//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Event client interface
EventsSDK mnuboEventClient = mnuboClient.getEventClient();

//read the event from the JSON file
String event2BeSent = readingFile( "myEvents.json" );

//deserialise the JSON file
EventValues event2Send = SDKMapperUtils.readValue( event2BeSent , Events.class );

//send the event
List<Result> results = mnuboEventClient.send( objectID, event2Send );
```

The JSON file " myEventsByObjectFile.json " is as follows:

```
[
    {
        "x_event_type" : "thermostat_temperature",
        "event_id": "11111111-2222-3333-4444-555555555555",
        "thermostat_temperature": 20,
        "errorcode": "",
        "varname": "temperature",
    },
    {
        "x_event_type" : "thermostat_temperature",
        "event_id": "11111111-2222-3333-6666-555555555555",
        "thermostat_temperature": 22,
        "cause_type": "normal",
        "varname": "temperature",
    }
]
```

You can verify if there is any error(s) for the list of owners created / updated.

The API result will contain a list of the “x_device_id”s (id), their results (“success” or “error”)
and the message if there is an error.

Note: ResultStates is an ENUM with only ResultStates.success (“success”) or ResultStates.error (“error”)

Here is an example to get the data of the list of Result:
```
for(Result r: results) {
    String id = r.getId();
    ResultStates resultState = r.getResult();
    String message = r.getMessage();
}
```

### To send multiple events to multiple objects:

```
//private String objectID = "mythermostat0301424";

//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Event client interface
EventsSDK mnuboEventClient = mnuboClient.getEventClient();

//build the events
List<Event> event2Send = new ArrayList<Event>();
Event event1 = Event.builder()
                    .withEventType("thermostat_temperature")
                    .withSmartObject("connect_alpha.6hv135nw00393.81")
                    .withTimestamp(DateTime.now())
                    .withAddedTimeseries("temperature", 20)
                    .withAddedTimeseries("errorcode", "")
                    .withAddedTimeseries("varname", "temperature")
                    .build();
event2Send.add(event1);
Event event2 = Event.builder()
                    .withEventType("light_dimmer.internal")
                    .withSmartObject("connect_alpha.6hv135nw00393.82")
                    .withEventID(UUID.fromString("d7284e6b-cf62-4d70-bf79-6ff81af97d7f"))
                    .withTimestamp(DateTime.now())
                    .withAddedTimeseries("light_dimmer", 0)
                    .withAddedTimeseries("cause_type", "internal")
                    .withAddedTimeseries("varname", "level")
                    .withAddedTimeseries("varfunction", "light-dimmer")
                    .build();
event2Send.add(event2);
Event event3 = Event.builder()
                    .withEventType("mask_masked.internal")
                    .withSmartObject("connect_alpha.6hv135nw00393.83")
                    .withTimestamp(DateTime.now())
                    .withAddedTimeseries("mask_masked", "Not Masked")
                    .withAddedTimeseries("cause_type", "internal")
                    .withAddedTimeseries("varname", "mask-state")
                    .build();
event2Send.add(event3);

//send the event
List<Result> results = mnuboEventClient.send( event2Send );
```

You can also send multiple events to multiple objects using a JSON file:

```
//get a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Event client interface
EventsSDK mnuboEventClient = mnuboClient.getEventClient();

//read event from the JSON file
String event2Besent = readingFile( "myEvents.json" );

//deserialise the JSON file
EventValues event2Send = SDKMapperUtils.readValue( event2Besent , EventValues.class );

//send the event
List<Result> results = mnuboEventClient.send( event2Send );
```

The JSON file "myEvents.json" is as follows:

```
[
    {
        "x_object":
        {
            "x_device_id" :"connect_alpha.6hv135nw00393.81"
        },
        "x_event_type": "thermostat_temperature",
        "event_id": "11111111-2222-3333-4444-555555555555",
        "thermostat_temperature": 20,
        "cause_type": null,
        "errorcode": "",
        "varname": "temperature",
    },
    {
        "x_object":
        {
            "x_device_id" :"connect_alpha.6hv135nw00393.82"
        },
        "x_event_type": "light_dimmer.internal",
        "event_id": "11111111-2222-3333-6666-555555555555",
        "light_dimmer": 0,
        "cause_type": "internal",
        "errorcode": null,
        "varname": "level",
        "varfunction": "light-dimmer",
    },
    {
        "x_object":
        {
            "x_device_id" :"connect_alpha.6hv135nw00393.83"
        },
        "x_event_type": "mask_masked.internal",
        "x_timestamp": "2014-10-30T21:46:10+00:00",
        "event_id": "11111111-2222-1111-6666-555555555555",
        "mask_masked": "Not Masked",
        "cause_type": "internal",
        "errorcode": null,
        "varname": "mask-state",
    }
]
```

You can verify if there is any error(s) for the list of events sent.

The API result will contain a list of the “event_id”s (id), their results (“success” or “error”)
and the message if there is an error.

Note: ResultStates is an ENUM with only ResultStates.success (“success”) or ResultStates.error (“error”)

Here is an example to get the data of the list of Result:
```
for(Result r: results) {
    String id = r.getId();
    ResultStates resultState = r.getResult();
    String message = r.getMessage();
}
```

Check if an event exists
------------------------
This example describes how to validate if an event exists:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Event client interface
EventsSDK mnuboEventClient = mnuboClient.getEventClient();

//true if the event exists, false if not.
boolean doesExists = mnuboEventClient.eventExists( UUID.fromString("11111111-2222-1111-6666-555555555555") );
```

Check if a batch of events exist
--------------------------------
This example describes how to validate if a list of events exists:
```
//Request a mnubo client using the basic method.
MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient( HOST , CONSUMER_KEY , CONSUMER_SECRET );

//get an Event client interface
EventsSDK mnuboEventClient = mnuboClient.getEventClient();

Map<String, Boolean> exists = mnuboEventClient.eventsExist( Arrays.asList(UUID.fromString("22222222-2222-2222-2222-222222222222"), UUID.fromString("11111111-1111-1111-1111-111111111111")) );

CHANGELOG.md
<a name="1.7.7"></a>

# [1.7.7](https://github.com/mnubo/mnubo-java-sdk/compare/v1.7.6...v1.7.7) (2016-07-25)

### Features

* **POST /api/v3/events:** support for 'conflict' and 'notfound' error codes

<a name="1.7.6"></a>

# [1.7.6](https://github.com/mnubo/mnubo-java-sdk/compare/v1.7.5...v1.7.6) (2016-07-21)

### Features

* **POST /api/v3/events:** this API does not allow anymore to overwrite an existing event
* **POST /api/v3/events/exists:** new endpoint to check the existence of a list of events
* **GET /api/v3/events/exists/{event_id}:** new endpoint to check the existence of a specific event
* **POST /api/v3/objects/exists:** new endpoint to check the existence of a list of objects
* **GET /api/v3/objects/exists/{device_id}:** new endpoint to check the existence of a specific object
* **POST /api/v3/owners/exists:** new endpoint to check the existence of a list of owners
* **GET /api/v3/owners/exists/{username}:** new endpoint to check the existence of a specific owner


example/src/main/java/client/SDKUsageExample.java
/*
 * ---------------------------------------------------------------------------
 *
 * COPYRIGHT (c) 2015 Mnubo Inc. All Rights Reserved.
 *
 * The copyright to the computer program(s) herein is the property of Mnubo Inc. The program(s) may be used and/or
 * copied only with the written permission from Mnubo Inc. or in accordance with the terms and conditions stipulated in
 * the agreement/contract under which the program(s) have been supplied.
 *
 * Author: marias Date : Jul 16, 2015
 *
 * ---------------------------------------------------------------------------
 */

package client;

import java.util.HashMap;
import java.util.List;
import java.util.UUID;

import org.joda.time.DateTime;

import com.mnubo.java.sdk.client.factory.MnuboSDKFactory;
import com.mnubo.java.sdk.client.models.Event;
import com.mnubo.java.sdk.client.models.Owner;
import com.mnubo.java.sdk.client.models.SmartObject;
import com.mnubo.java.sdk.client.spi.MnuboSDKClient;

public class SDKUsageExample {
    private static String HOST = "here the host, example 'rest.sandbox.mnubo.com'";
    private static String CONSUMER_KEY = "here you consumer key";
    private static String CONSUMER_SECRET = "here you consumer secret";

    private MnuboSDKClient mnuboClient = MnuboSDKFactory.getClient(HOST, CONSUMER_KEY, CONSUMER_SECRET);

    public SDKUsageExample() throws Exception {
    }

    /**
     *
     * This is an example to create a new object or update it if already exists.
     *
     * @param object2Create, Object to be created.
     */
    public void createObject(SmartObject object2Create) {
        // Posting object SDK call
        mnuboClient.getObjectClient().create(object2Create);
    }

    /**
     *
     * This is an example to post event to ONE object.
     *
     * @param objectId, Object to be posted,
     * @param events, events to be posted in the object. Please check resource
     * eventvalues.json file to get an example about the this object.
     */
    public void postEvents2Object(String objectId, List<Event> events) {
        // Posting events to only one object
        mnuboClient.getEventClient().send(objectId, events);
    }

    /**
     *
     * This is an example to post event to ONE or SEVERAL objects.
     *
     * @param events, events to be posted in objects. Please check resource event.json
     * file to get an example about the this object.
     */
    public void postEvents2Object(List<Event> events) {
        // Posting events to only one object
        mnuboClient.getEventClient().send(events);
    }

    /**
     *
     * This is an example to create a new owner.
     *
     * @param owner2Create, Owner to be created.
     */
    public void createOwner(Owner owner2Create) {
        // Posting events to only one object
        mnuboClient.getOwnerClient().create(owner2Create);
    }

    /**
     *
     * This is an example how to build a new owner. Please note that only "username" is a
     * mandatory field, others ones are optional.
     *
     */
    public Owner buildingOwner() {
        //@formatter:off
        return Owner.builder()
                .withUsername(/* username, this is MANDATORY */"John@mnubotest.com")
                .withPassword(/* password */"PassWORD!??%3")
                .withRegistrationDate(/* datetime of registration */DateTime.parse("2014-01-27T08:01:01.000Z"))
                .withAttributes(/* add here all attributes as a MAP */new HashMap<String, Object>())
                .withAddedAttribute("Here the attribute key", "here its value as \"Object\"")
                .withAddedAttribute("age", 35)
                .withAddedAttribute("contry", "USA")
                .build();
        //@formatter:on
    }

    /**
     *
     * This is an example how to build a new SmartObject. Please note that "deviceId" and
     * "ObjectType" are mandatory fields, others ones are optional.
     *
     */
    public SmartObject buildingSmartObject() {
        //@formatter:off
        return SmartObject.builder()
                .withDeviceId(/* deviceId, this is MANDATORY */"Room_temperature_02154")
                .withObjectType(/* object type, this is MANDATORY */"Thermostat")
                .withOwner(/* owner's username */"John@mnubotest.com")
                .withRegistrationDate(/* datetime of registration */DateTime.now())
                .withAttributes(/* add here all attributes as a MAP */new HashMap<String, Object>())
                .withAddedAttribute("Here the attribute key", "here its value as \"Object\"")
                .withAddedAttribute("contry", "Sydney")
                .withAddedAttribute("address", "42 Wallaby Way")
                .withAddedAttribute("film", "nemo")
                .build();
        //@formatter:on
    }

    /**
     *
     * This is an example how to build a new Event. Please note that "deviceId" and
     * "ObjectType" are mandatory fields, others ones are optional.
     *
     */
    public Event buildingEvent() {
        //@formatter:off
       return Event.builder()
               .withEventType(/* eventType, this is MANDATORY */"temperature")
               .withEventID(/* event id */ UUID.fromString("11111111-2222-3333-4444-555555555555"))
               .withSmartObject(/* object's deviceId */"Room_temperature_02154")
               .withTimestamp(/* datetime of posting */DateTime.now())
               .withTimeseries(/* add here all attributes as a MAP */new HashMap<String, Object>())
               .withAddedTimeseries("Here the attribute key", "here its value as \"Object\"")
               .withAddedTimeseries("temperature", 75)
               .withAddedTimeseries("id", "0238510.5654.56")
               .build();
       //@formatter:on
    }
}
